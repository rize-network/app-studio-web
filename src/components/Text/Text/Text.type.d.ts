export type TextWeights =
  | 'hairline'
  | 'thin'
  | 'light'
  | 'normal'
  | 'medium'
  | 'semiBold'
  | 'bold'
  | 'extraBold'
  | 'black';

export type Size =
  | 'xs'
  | 'sm'
  | 'md'
  | 'lg'
  | 'xl'
  | '2xl'
  | '3xl'
  | '4xl'
  | '5xl'
  | '6xl';

export type Headings = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
