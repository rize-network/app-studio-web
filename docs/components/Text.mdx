### **Import**

```tsx
import { Text } from "@app-studio/web";
```

### **Default**

```tsx
<Text>Some text here</Text>
```

### **Sizes**

“**_size_**” changes the text-size. It has a type "Sizes" with default value “md”.

```tsx
import { Vertical } from "../Layout/Vertical/Vertical";

<Vertical gap={5} alignItems="center">
  {["xs", "sm", "md", "lg", "xl", "xl", "xl", "xl", "xl", "6xl"].map(
    (size, index) => (
      <Text key={index} size={size}>
        {size}
      </Text>
    )
  )}
</Vertical>;
```

### **Headings**

“**_heading_**” renders a specific heading html tag. “h1” indicates the most important heading and “h6” the least important heading.

```tsx
import { Vertical } from "../Layout/Vertical/Vertical";
<Vertical gap={2}>
  <Text heading="h1">Heading 1</Text>
  <Text heading="h2">Heading 2</Text>
  <Text heading="h3">Heading 3</Text>
  <Text heading="h4">Heading 4</Text>
  <Text heading="h5">Heading 5</Text>
  <Text heading="h6">Heading 6</Text>
</Vertical>;
```

### **Strike**

“**_isStriked_**” marks up a text to indicate that it is no longer valid.

```tsx
<Text color="theme.secondary" isStriked>
  Some text here
</Text>
```

### **Subscript**

“**_isSub_**” makes text appears slightly below the baseline of the surrounding text.

```tsx
<Text>
  H<Text isSub>2</Text>0
</Text>
```

### **Superscript**

“isSup” makes text appears slightly above the surrounding text.

```tsx
<Text>
  H<Text isSup>2</Text>0
</Text>
```

### **Weights**

“**_weight_**” changes the font weight of the text. It have a default value of “medium”.

```tsx
import { Vertical } from "../Layout/Vertical/Vertical";

<Vertical gap={5} alignItems="center">
  {[
    "hairline",
    "thin",
    "light",
    "normal",
    "medium",
    "semiBold",
    "bold",
    "extraBold",
    "black",
  ].map((weight, index) => (
    <Text key={index} weight={weight}>
      {weight}
    </Text>
  ))}
</Vertical>;
```

### **Truncate**

“**__**” removes a part of the text and add an ellipsis to the end.

```tsx
import { Center } from "app-studio";

<Center>
  <Text maxLines={3} width="200px" >
    Pellentesque habitant morbi tristique senectus et netus et malesuada fames
    ac turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget,
    tempor sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean
    ultricies mi vitae est.
  </Text>
</Center>;
```

### **Italic**

“**_isItalic_**” styles the text in italic.

```tsx
<Text isItalic>Some text here</Text>
```

### **Underline**

“**_isUnderlined_**” underlines the text.

```tsx
<Text isUnderlined>Some text here</Text>
```

## **Types**

```tsx
export type Sizes =
  | "xs"
  | "sm"
  | "md"
  | "lg"
  | "xl"
  | "xl"
  | "xl"
  | "xl"
  | "xl"
  | "6xl";
```

```tsx
type TextWeights =
  | "hairline"
  | "thin"
  | "light"
  | "normal"
  | "medium"
  | "semibold"
  | "bold"
  | "extrabold"
  | "black";
```
