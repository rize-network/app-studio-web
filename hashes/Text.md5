6950792681612697206dfc3807969365