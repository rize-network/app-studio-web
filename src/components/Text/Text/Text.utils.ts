const normalizeHexColor = (backgroundColor: string) => {
  if (!backgroundColor) {
    return null;
  }

  const trimmed = backgroundColor.trim();
  const withoutHash = trimmed.startsWith('#') ? trimmed.slice(1) : trimmed;

  if (withoutHash.length === 3) {
    if (!/^[0-9a-fA-F]{3}$/.test(withoutHash)) {
      return null;
    }
    return `#${withoutHash
      .split('')
      .map((char) => char + char)
      .join('')}`;
  }

  if (!/^[0-9a-fA-F]{6}$/.test(withoutHash)) {
    return null;
  }

  return `#${withoutHash.toLowerCase()}`;
};

export const getTextColorHex = (backgroundColor: string) => {
  const normalized = normalizeHexColor(backgroundColor);

  if (!normalized) {
    return 'black';
  }

  // Simple luminance calculation to determine text color contrast
  const color = normalized.replace('#', '');
  const r = parseInt(color.substring(0, 2), 16);
  const g = parseInt(color.substring(2, 4), 16);
  const b = parseInt(color.substring(4, 6), 16);
  const luminance = (0.299 * r + 0.587 * g + 0.114 * b) / 255;
  return luminance > 0.4 ? 'black' : 'white';
};

export const getTextColor = (backgroundColor: string) =>
  getTextColorHex(backgroundColor);
