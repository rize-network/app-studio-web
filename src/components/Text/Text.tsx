import React from 'react';

import { TextProps } from './Text/Text.props';
import TextView from './Text/Text.view';

const TextComponent: React.FC<TextProps> = (props) => {
  return <TextView {...props} />;
};

/**
 * The Text component is a simple component that renders a text string or paragraphs as a DOM element in the UI. It is a <p> tag by default.
 */
export const Text = TextComponent;
