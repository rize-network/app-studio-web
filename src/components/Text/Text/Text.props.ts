import { CSSProperties } from 'react';
import type { TextProps as TextAppProps } from 'app-studio';
import { Headings, Size, TextWeights } from './Text.type';
export interface TextProps
  extends Omit<TextAppProps, 'children' | 'style' | 'pointerEvents'>,
    CSSProperties {
  children: React.ReactNode;
  heading?: Headings;
  isItalic?: boolean;
  isStriked?: boolean;
  isUnderlined?: boolean;
  isSub?: boolean;
  isSup?: boolean;
  maxLines?: number;
  size?: Size;
  weight?: TextWeights;
}
export interface ContentProps {
  children: React.ReactNode | string;
  isSub?: boolean;
  isSup?: boolean;
}
export interface TruncateTextProps {
  text: string;
  maxLines?: number;
}
