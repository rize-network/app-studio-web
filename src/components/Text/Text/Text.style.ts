/**
 * Text Styles
 *
 * Defines the styles for the Text component following the design guidelines:
 * - Typography: Inter/Geist font, specific sizes/weights
 * - Spacing: 4px grid system
 * - Colors: Neutral palette with semantic colors
 */

import { Headings, Size, TextWeights } from './Text.type';
import { ViewProps } from 'app-studio';

/**
 * Heading sizes following typography guidelines
 */
export const HeadingSizes: Record<Headings, ViewProps> = {
  h1: {
    fontSize: '36px', // 2.25rem
    lineHeight: '40px', // 2.5rem
    fontWeight: '700', // Bold
    letterSpacing: '-0.02em',
    marginBottom: '24px', // 6 × 4px grid
  },
  h2: {
    fontSize: '30px', // 1.875rem
    lineHeight: '36px', // 2.25rem
    fontWeight: '700', // Bold
    letterSpacing: '-0.02em',
    marginBottom: '20px', // 5 × 4px grid
  },
  h3: {
    fontSize: '24px', // 1.5rem
    lineHeight: '32px', // 2rem
    fontWeight: '600', // Semi-bold
    letterSpacing: '-0.01em',
    marginBottom: '16px', // 4 × 4px grid
  },
  h4: {
    fontSize: '20px', // 1.25rem
    lineHeight: '28px', // 1.75rem
    fontWeight: '600', // Semi-bold
    letterSpacing: '-0.01em',
    marginBottom: '16px', // 4 × 4px grid
  },
  h5: {
    fontSize: '18px', // 1.125rem
    lineHeight: '24px', // 1.5rem
    fontWeight: '600', // Semi-bold
    letterSpacing: '-0.01em',
    marginBottom: '12px', // 3 × 4px grid
  },
  h6: {
    fontSize: '16px', // 1rem
    lineHeight: '24px', // 1.5rem
    fontWeight: '600', // Semi-bold
    letterSpacing: '-0.01em',
    marginBottom: '8px', // 2 × 4px grid
  },
};

/**
 * Font sizes following typography guidelines
 */
export const FontSizes: Record<Size, string> = {
  xs: '12px', // 0.75rem
  sm: '14px', // 0.875rem
  md: '16px', // 1rem (base)
  lg: '18px', // 1.125rem
  xl: '20px', // 1.25rem
};

/**
 * Line heights following typography guidelines
 */
export const LineHeights: Record<Size, string> = {
  xs: '16px', // 1rem
  sm: '20px', // 1.25rem
  md: '24px', // 1.5rem
  lg: '28px', // 1.75rem
  xl: '28px', // 1.75rem
};

/**
 * Font weights following typography guidelines
 */
export const FontWeights: Record<TextWeights, string> = {
  hairline: '100',
  thin: '200',
  light: '300',
  normal: '400',
  medium: '500',
  semiBold: '600',
  bold: '700',
  extraBold: '800',
  black: '900',
};
