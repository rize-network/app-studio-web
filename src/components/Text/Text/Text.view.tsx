/**
 * Text View Component
 *
 * Renders text with various styles and states according to the design guidelines.
 */

import React, { useEffect, useRef, useState } from 'react';
import { Element, View, ViewProps } from 'app-studio';

import { TextProps } from './Text.props';
import {
  HeadingSizes,
  FontSizes,
  LineHeights,
  FontWeights,
} from './Text.style';

interface Props extends TextProps {
  views?: {
    container?: ViewProps;
  };
}

interface ContentProps extends React.HTMLAttributes<HTMLElement> {
  children: React.ReactNode;
  isSub?: boolean;
  isSup?: boolean;
  views?: {
    sup?: ViewProps;
  };
}

interface TruncateTextProps extends React.HTMLAttributes<HTMLDivElement> {
  text: string;
  maxLines?: number;
  views?: {
    truncateText?: ViewProps;
  };
}

/**
 * Renders text content with support for subscript and superscript
 */
const TextContent: React.FC<ContentProps> = ({
  children,
  isSub,
  isSup,
  views,
}) => (
  <>
    {typeof children === 'string' ? (
      <>
        {isSub && (
          <View
            as="sub"
            fontSize="75%"
            lineHeight="0"
            position="relative"
            bottom="-0.25em"
            {...views?.sup}
          >
            {children}
          </View>
        )}
        {isSup && (
          <View
            as="sup"
            fontSize="75%"
            lineHeight="0"
            position="relative"
            top="-0.5em"
            {...views?.sup}
          >
            {children}
          </View>
        )}
        {!isSub && !isSup && <>{children}</>}
      </>
    ) : (
      children
    )}
  </>
);

/**
 * Renders text with truncation after a specified number of lines
 */
const TruncateText: React.FC<TruncateTextProps> = ({
  text,
  maxLines = 1,
  views,
}) => {
  const containerRef = useRef<HTMLDivElement>(null);
  const [truncatedLength, setTruncatedLength] = useState(text.length);

  useEffect(() => {
    const textNode = containerRef.current;
    if (!textNode) return;

    const updateTruncatedText = () => {
      const comLineHeight = getComputedStyle(textNode).lineHeight;
      const lineHeight =
        comLineHeight !== 'normal' ? parseFloat(comLineHeight) : 20;
      const maxHeight = lineHeight * maxLines;

      let start = 0;
      let end = text.length;
      let middle;

      while (start <= end) {
        middle = Math.floor((start + end) / 2);
        textNode.innerText = text.substring(0, middle) + '...';
        const currentHeight = textNode.offsetHeight;

        if (currentHeight > maxHeight) {
          end = middle - 1;
        } else {
          start = middle + 1;
        }
      }

      setTruncatedLength(end);
    };

    updateTruncatedText();
  }, [text, maxLines]);

  const displayText =
    text.length > truncatedLength
      ? text.substring(0, truncatedLength) + '...'
      : text;

  return (
    <View
      ref={containerRef}
      overflow="hidden"
      textOverflow="ellipsis"
      {...views?.truncateText}
    >
      {displayText}
    </View>
  );
};

/**
 * Main Text component that renders text with various styles and states
 */
const TextView: React.FC<Props> = ({
  children,
  heading,
  maxLines,
  isItalic = false,
  isUnderlined = false,
  isSub = false,
  isSup = false,
  isStriked = false,
  weight = 'normal',
  size = 'md',
  views,
  ...props
}) => {
  // Apply heading styles if a heading is specified
  const headingStyles = heading ? HeadingSizes[heading] : {};

  // For sub/sup text, use inline display
  const noLineBreak = isSub || isSup ? { display: 'inline' } : {};

  // Get font size, line height, and weight from our design system
  const fontSize = FontSizes[size];
  const lineHeight = LineHeights[size];
  const fontWeight = FontWeights[weight];

  return (
    <Element
      // Apply typography styles according to design guidelines
      fontSize={fontSize}
      lineHeight={lineHeight}
      fontStyle={isItalic ? 'italic' : 'normal'}
      fontWeight={fontWeight}
      letterSpacing="-0.01em"
      textDecoration={
        isStriked ? 'line-through' : isUnderlined ? 'underline' : 'none'
      }
      color="color.gray.900"
      // Apply dark mode styles
      _dark={{
        color: 'color.gray.100',
      }}
      // Apply layout styles
      {...noLineBreak}
      // Apply heading styles if specified
      {...headingStyles}
      // Apply any custom props
      {...props}
      // Apply custom container styles
      {...views?.container}
    >
      {maxLines && typeof children === 'string' ? (
        <TruncateText text={children} maxLines={maxLines} />
      ) : (
        <TextContent isSub={isSub} isSup={isSup}>
          {children}
        </TextContent>
      )}
    </Element>
  );
};

export default TextView;
