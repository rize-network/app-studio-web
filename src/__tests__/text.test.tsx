import React from 'react';
import renderer from 'react-test-renderer';
import { Text } from 'src/components/Text/Text';
import { Size } from 'src/components/Text/Text/Text.type';

import { cleanup, render, screen } from '@testing-library/react';

afterEach(() => {
  cleanup();
});

test('should render text component without crashing', () => {
  render(<Text>Some Text</Text>);
});

test('should render text component with the correct text', () => {
  render(<Text>Some text here</Text>);
  const textElement = screen.getByRole('text');
  expect(textElement).toBeInTheDocument();
  expect(textElement).toHaveTextContent('Some text here');
});

test('should render text component with correct class name', () => {
  render(<Text className="text">Some text here</Text>);
  const textElement = screen.getByRole('text');
  expect(textElement).toHaveClass('text');
});

test('should render a text with fontSize equals to sm', () => {
  render(<Text size={'sm' as Size}>Some text here</Text>);
  const textElement = screen.getByRole('text');
  expect(textElement).toHaveStyle({ fontSize: 'sm' });
});

test('should render a text with a weight equals to bold', () => {
  render(<Text weight="bold">Some text here</Text>);
  const textElement = screen.getByRole('text');
  expect(textElement).toHaveStyle({ fontWeight: 700 });
});

test('should render a text with an italic style', () => {
  render(<Text isItalic>Some text here</Text>);
  const textElement = screen.getByRole('text');
  expect(textElement).toHaveStyle({ fontStyle: 'italic' });
});

test('should render a subscript text to match snapshot', () => {
  const tree = renderer
    .create(
      <Text>
        H<Text isSub>2</Text>O
      </Text>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('should render a superscript text to match snapshot', () => {
  const tree = renderer
    .create(
      <Text>
        H<Text isSup>2</Text>O
      </Text>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('should render one truncated text line to match snapshot', () => {
  const tree = renderer
    .create(
      <Text isUnderlined maxLines={2} isTruncated>
        Pellentesque habitant morbi tristique senectus et netus et malesuada
        fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae,
        ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam
        egestas semper. Aenean ultricies mi vitae est. Mauris placerat eleifend
        leo.
      </Text>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('text to match snapshot', () => {
  const tree = renderer
    .create(
      <Text heading="h1" weight="bold" isStriked>
        Some text here
      </Text>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
