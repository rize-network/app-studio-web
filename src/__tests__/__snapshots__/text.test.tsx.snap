// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a subscript text to match snapshot 1`] = `
<div
  className="sc-bdfCDU FxXNY"
  fontSize={14}
  fontStyle="normal"
  fontWeight={400}
  role="text"
  textDecoration="none"
>
  H
  <div
    className="sc-bdfCDU hVEnQK"
    display="inline"
    fontSize={14}
    fontStyle="normal"
    fontWeight={400}
    role="text"
    textDecoration="none"
  >
    <sup>
      2
    </sup>
  </div>
  O
</div>
`;

exports[`should render a superscript text to match snapshot 1`] = `
<div
  className="sc-bdfCDU FxXNY"
  fontSize={14}
  fontStyle="normal"
  fontWeight={400}
  role="text"
  textDecoration="none"
>
  H
  <div
    className="sc-bdfCDU hVEnQK"
    display="inline"
    fontSize={14}
    fontStyle="normal"
    fontWeight={400}
    role="text"
    textDecoration="none"
  >
    <sup>
      2
    </sup>
  </div>
  O
</div>
`;

exports[`should render one truncated text line to match snapshot 1`] = `
<div
  className="sc-bdfCDU hXOWac"
  fontSize={14}
  fontStyle="normal"
  fontWeight={400}
  role="text"
  textDecoration="underline"
>
  <div>
    Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies mi vitae est. Mauris placerat eleifend leo.
  </div>
</div>
`;

exports[`text to match snapshot 1`] = `
<div
  className="sc-bdfCDU dxpfnQ"
  fontSize={96}
  fontStyle="normal"
  fontWeight={700}
  letterSpacing={-1.5}
  role="text"
  textDecoration="line-through"
>
  Some text here
</div>
`;
